compiler/rustc_mir/src/const_eval/machine.rs

```diff
@@ -237,10 +252,14 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
             if !ecx.tcx.is_const_fn_raw(def.did) {
                 // Some functions we support even if they are non-const -- but avoid testing
                 // that for const fn!
-                ecx.hook_panic_fn(instance, args)?;
-                // We certainly do *not* want to actually call the fn
-                // though, so be sure we return here.
-                throw_unsup_format!("calling non-const function `{}`", instance)
+                if let Some(new_instance) = ecx.hook_panic_fn(instance, args)? {
+                    // We call another const fn instead.
```

Comment:
This is not checking that that function is actually a `const fn`. To be on the safe side, should we maybe add an assertion here or just before `load_mir`?

Reply:
I replaced it with a recursive call to make sure the same check gets applied to the new instance too.